luaext/lposix.c

```diff
@@ -330,26 +330,64 @@ static int Pmkfifo(lua_State *L)		/** mkfifo(path) */
 }
 
 
+static void set_cloexec(int fd)
+{
+	int flag = fcntl(fd, F_GETFD);
+
+	if (flag == -1 || (flag & FD_CLOEXEC))
+		return;
+
+	fcntl(fd, F_SETFD, FD_CLOEXEC);
+}
+
+static void set_cloexec_all(void)
+{
+	const int min_fd = 3; // don't touch stdin/out/err
+	const int default_open_max = 1024;
+	int fd, open_max;
+
+	open_max = sysconf(_SC_OPEN_MAX);
+	if (open_max == -1) {
+	    open_max = default_open_max;
+	    goto fallback;
+	}
+
+	if (open_max <= default_open_max) {
+		goto fallback;
+	}
+
+	// iterate over fds obtained from /proc
+	DIR *dir = opendir("/proc/self/fd");
+	if (dir == NULL) {
+		goto fallback;
+	}
+
+	struct dirent *entry;
+	while ((entry = readdir(dir)) != NULL) {
+		fd = atoi(entry->d_name);
+		if (fd >= min_fd)
+			set_cloexec(fd);
+	}
+	closedir(dir);
+
+	return;
+
+fallback:
+	// iterate over all possible fds
+	for (fd = min_fd; fd < open_max; fd++)
+		set_cloexec(fd);
+}
+
 static int Pexec(lua_State *L)			/** exec(path,[args]) */
```

Comment:
You need to do the same change for doScriptExec() in lib/rpmscript.c as well. At least in our case, that is where all the time is spent.